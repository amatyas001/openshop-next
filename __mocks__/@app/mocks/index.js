export const mockProduct = (repeat = 1) => {
  let index = 0;
  let items = [];
  do {
    items.push({
      id: `mock_id_${index}`,
      amount: 1,
      color: `mock_color_${index}`,
      description: `mock_description_${index}`,
      gender: `mock_gender_${index}`,
      img: `mock_img_${index}`,
      name: `mock_name_${index}`,
      price: 1,
      review: `mock_review_${index}`,
      sizes: [`mock_size_0_${index}`, `mock_size_1_${index}`],
      starrating: 1,
    });
  } while (--repeat !== 0);
  return items;
};
