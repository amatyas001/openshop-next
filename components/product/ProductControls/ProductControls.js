import PropTypes from 'prop-types';
import { FaStar } from 'react-icons/fa';
import { Heading, Text, Flex, Box, SimpleGrid } from '@chakra-ui/core';
import * as COLORS from '@app/config/colors';
import {
  ProductControlsAmount,
  ProductControlsButton,
  ProductControlsColor,
  ProductControlsSize,
} from '@app/components';

/**
 * Wrapper for input elements of a specific product buying details.
 *
 * @see https://amatyas001.github.io/openshop-next/#productcard
 */
export const ProductControls = (props) => {
  const { product } = props;
  const [details, setDetails] = React.useState({
    ...product,
    buy: { amount: 0 },
  });

  return (
    <SimpleGrid
      alignItems='center'
      columns='2'
      width={{ sm: '100%', xl: '30%' }}
      minWidth={{ xl: '400px' }}
    >
      <span />
      <Heading
        alignSelf='center'
        fontSize='2rem'
        m='0'
        ml='auto'
        textAlign='center'
        color={COLORS.HEADING.dark}
      >
        {`${product.price}&nbsp;$`}
      </Heading>
      <ProductControlsColor
        product={product}
        details={details}
        setDetails={
          /* istanbul ignore next */
          (e) =>
            setDetails((prev) => ({ ...prev, buy: { ...prev.buy, color: e } }))
        }
      />
      <Flex fontSize='1.4rem' ml='auto' mb='2px' {...props}>
        {[...Array(Math.floor(product.starrating))].map((e, i) => (
          /* eslint react/no-array-index-key: "off" */
          <Box key={i} as={FaStar} color={COLORS.HEADING.light} />
        ))}
      </Flex>
      <ProductControlsSize
        product={product}
        details={details}
        setDetails={
          /* istanbul ignore next */
          (e) =>
            setDetails((prev) => ({ ...prev, buy: { ...prev.buy, size: e } }))
        }
      />
      <Text
        color={COLORS.BUTTON.primary.base}
        fontFamily='Khand'
        fontSize='1.4rem'
        fontWeight='bold'
        ml='auto'
        textTransform='uppercase'
      >
        {`${product.amount} in stock`}
      </Text>
      <ProductControlsAmount
        product={product}
        details={details}
        setDetails={
          /* istanbul ignore next */
          (e) =>
            setDetails((prev) => ({ ...prev, buy: { ...prev.buy, amount: e } }))
        }
      />
      <ProductControlsButton product={details} setDetails={setDetails} />
    </SimpleGrid>
  );
};

ProductControls.propTypes = {
  /**
   * Unique product which controls are rendered
   */
  product: PropTypes.shape({
    id: PropTypes.string.isRequired,
    amount: PropTypes.number.isRequired,
    color: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.arrayOf(PropTypes.string),
    ]),
    description: PropTypes.string,
    gender: PropTypes.string,
    img: PropTypes.string,
    name: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    review: PropTypes.string,
    sizes: PropTypes.arrayOf(PropTypes.string),
    starrating: PropTypes.number.isRequired,
  }).isRequired,
};
