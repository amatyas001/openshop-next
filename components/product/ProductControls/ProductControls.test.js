import { create, act } from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { mockProduct } from '@app/mocks';
import { ProductControls } from './ProductControls';

const mockStore = configureStore([]);

describe('<ProductControls />', () => {
  let tree, store;

  const product = mockProduct(1)[0];

  beforeAll(() => {
    store = mockStore({});

    act(() => {
      tree = create(
        <Provider store={store}>
          <ProductControls product={product} />
        </Provider>
      );
    });
  });

  it('should render with required props', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
