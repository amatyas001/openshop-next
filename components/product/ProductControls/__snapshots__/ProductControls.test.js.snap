// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductControls /> should render with required props 1`] = `
<div
  className="css-1j174qx"
>
  <span />
  <h2
    className="css-d94gct"
  >
    1
     $
  </h2>
  <p
    className="css-u84yc3"
  >
    mock_color_0
  </p>
  <div
    className="css-1nqhd46"
  >
    <svg
      className="css-1qla8s8"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 576 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
      />
    </svg>
  </div>
  <div
    className="css-1php3x8"
  >
    <select
      className="css-1fbi3l4"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        Select size...
      </option>
      <option
        value="mock_size_0_0"
      >
        mock_size_0_0
      </option>
      <option
        value="mock_size_1_0"
      >
        mock_size_1_0
      </option>
    </select>
    <div
      className="css-rl6gbg"
    >
      <svg
        aria-hidden="true"
        className="css-k176ff"
        data-custom-icon={true}
        fill="currentColor"
        focusable="false"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
        />
      </svg>
    </div>
  </div>
  <p
    className="css-1nsgbpp"
  >
    1
     in stock
  </p>
  <div
    className="css-1dm3qsp"
    data-testid="product-amount-wrapper"
  >
    <div
      className="css-hboir5"
    >
      <div
        aria-hidden={true}
        className="css-1yqqx3k"
      >
        <button
          className="css-fdetwv"
          data-testid="product-amount-increment"
          href="#void"
          onClick={null}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          role="button"
          tabIndex="0"
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-32 252c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92H92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
            />
          </svg>
        </button>
        <button
          className="css-1fd1i4r"
          data-testid="product-amount-decrement"
          href="#void"
          onClick={null}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          role="button"
          tabIndex="0"
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zM92 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H92z"
            />
          </svg>
        </button>
      </div>
      <input
        aria-invalid={false}
        aria-readonly={true}
        aria-valuemax={1}
        aria-valuemin={0}
        aria-valuenow={0}
        autoComplete="off"
        className="css-s3c5ne"
        data-testid="product-amount-input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        readOnly={true}
        role="spinbutton"
        type="text"
        value={0}
      />
    </div>
  </div>
  <button
    className="css-1e5lh77"
    data-testid="product-buy-button"
    disabled={true}
    href="#void"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex="0"
    type="button"
  >
    buy
  </button>
</div>
`;
